### ISSUE TITLE
Monadic Mealy and Moore machines

### ISSUE THREAD

**Issue by treeowl** [ekmett/machines#51]
I think these probably deserve to exist. Here's how it goes for Moore machines (I didn't try to figure out a `Monad` instance since even if it exists it seems unlikely to be that useful):

### Moore
```
class AutomatonM f where
    autoM :: Monad m => f m a b -> ProcessT m a b
    
data Moorish m a b = Moorish b (a -> m (Moorish m a b))

instance AutomatonM Moorish where
    autoM = construct . moorishToPlanT
                   
moorishToPlanT :: Monad m => Moorish m a b -> PlanT (Is a) b m q
moorishToPlanT (Moorish b f) =
    do
      yield b
      a <- await
      foo <- lift $ f a
      moorishToPlanT foo

-- TODO This can be derived in GHC 7.10, yielding a Functor m constraint
instance Monad m => Functor (Moorish m a) where
    fmap f (Moorish b step) = Moorish (f b) (liftM (fmap f) . step)

instance Monad m => Applicative (Moorish m a) where
    pure x = blah
        where
          blah = Moorish x (\_ -> return blah)

    Moorish f stepf <*> Moorish x stepx =
        Moorish (f x) (\a -> ((<*>) `liftM` stepf a) `ap` stepx a)

instance Monad m => Profunctor (Moorish m) where
    rmap = fmap
    lmap f (Moorish x step) = Moorish x (liftM (lmap f) . step . f)

instance Monad m => Comonad (Moorish m a) where
    extract = copoint
    extend f w@(Moorish _ step) = Moorish (f w) (\i -> step i >>= return . extend f)

instance Monad m => ComonadApply (Moorish m a) where
    (<@>) = (<*>)

instance Monad m => Pointed (Moorish m a) where
    point a = loop where loop = Moorish a (const . return $ loop)

instance Copointed (Moorish m a) where
    copoint (Moorish x _) = x
```

**Comment by treeowl** [comment]
Actually, this probably doesn't go far enough. This extends `Process`-like functionality to `ProcessT`-like functionality, but we should be able to do better, allowing more general `MachineT`-like things, only with more constrained timing.

**Comment by acowley** [comment]
I'd like to see this for a `MealyM`, too. Is there any objection to adding them perhaps without all attendant instances?

**Comment by sordina** [comment]
I've got a use-case for MealyM, and have been chatting to @acowley about this in the concurrent-machines repo. I've sketched out an implementation here:

https://github.com/sordina/uniqhash/blob/master/Data/Machine/MealyM.hs

Let me know if you'd like a pull request for this newtype and associated classes.

**Comment by bens** [comment]
`MealyM` is [`MStreamF`](https://github.com/ivanperez-keera/dunai/blob/develop/src/Data/MonadicStreamFunction/Core.hs#L35) from `dunai`, which has an [paper](http://eprints.nottingham.ac.uk/36159/1/paper.pdf) about it.

**Comment by sordina** [comment]
@bens Ah so it is!

**Comment by acowley** [comment]
Thanks @bens, it's a good note. I'm not sure, though, that the existence of `MStreamF` should preclude adding `MealyM` to `machines`. Is there anything in `dunai` that would suggest an approach other than the obvious extension of `Mealy` to `MealyM`?

**Comment by bens** [comment]
Right, I wasn't suggesting MealyM shouldn't be included in `machines`, only that it would be worth looking at for interesting combinators and not as obvious instances.  Possibly `dunai` could have an instance for `AutoM` with a dependency on `machines`, but it's probably nicer for `machines` to be self contained.  Just some food for thought.

**Comment by sordina** [comment]
The other question that arrises from MealyM (and MooreM) is if Mealy should just be a type synonym for MealyM Identity...

**Comment by sordina** [comment]
Although "upgrade" should provide a good compromise: https://github.com/sordina/uniqhash/blob/master/Data/Machine/MealyM.hs#L67

**Comment by HuwCampbell** [comment]
```haskell
